Mixpanel/MixpanelExceptionHandler.m

```diff
@@ -69,22 +70,36 @@ - (instancetype)init {
         // Create a hash table of weak pointers to mixpanel instances
         _mixpanelInstances = [NSHashTable weakObjectsHashTable];
         
-        // Save the existing exception handler
-        _defaultExceptionHandler = NSGetUncaughtExceptionHandler();
+        _prev_signal_handlers = calloc(NSIG, sizeof(struct sigaction));
+
         // Install our handler
         [self setupHandlers];
     }
     return self;
 }
 
+- (void)dealloc {
+    free(_prev_signal_handlers);
+}
+
 - (void)setupHandlers {
+    _defaultExceptionHandler = NSGetUncaughtExceptionHandler();
     NSSetUncaughtExceptionHandler(&MPHandleException);
-    signal(SIGABRT, MPSignalHandler);
-    signal(SIGILL, MPSignalHandler);
-    signal(SIGSEGV, MPSignalHandler);
-    signal(SIGFPE, MPSignalHandler);
-    signal(SIGBUS, MPSignalHandler);
-    signal(SIGPIPE, MPSignalHandler);
+
+    struct sigaction action;
+    sigemptyset(&action.sa_mask);
+    action.sa_flags = SA_SIGINFO;
+    action.sa_sigaction = &MPSignalHandler;
+    int signals[] = {SIGABRT, SIGILL, SIGSEGV, SIGFPE, SIGBUS, SIGPIPE};
+    for (int i = 0; i < sizeof(signals) / sizeof(int); i++) {
+        struct sigaction prev_action;
+        int err = sigaction(signals[i], &action, &prev_action);
+        if (err == 0) {
+            memcpy(_prev_signal_handlers + signals[i], &prev_action, sizeof(prev_action));
+        } else {
+            NSLog(@"Errored while trying to set up sigaction for signal %d", signals[i]);
```

Comment:
can this actually happen? what cases would a setup of a known signal fail?

Reply:
i'm not exactly sure why it happens but PLCrashReporter handled this case https://opensource.plausible.coop/src/projects/PLCR/repos/plcrashreporter/browse/Source/PLCrashSignalHandler.mm#368